### ISSUE TITLE
Speedup of sendFrame method

### ISSUE THREAD

**Issue by cansik** [brunchboy/wayang#3]
I am currently using your code for connecting to the push. But the frame rate is not that high (about **20 FPS**). I think there are some speedup tweaks todo:

First of all, is there a reason why to just send 8 lines per transfer? Because the synchronous transfer is very time consuming. I just set the lines per transfer to 160 so we only have to send out the header and the data. Event he Ableton Push documentation does mention it: 

_"... the pixel data is typically sent using larger buffers, for example of 16kbytes each."_ ([doc](https://github.com/Ableton/push-interface/blob/master/doc/AbletonPush2MIDIDisplayInterface.asc#322-pixel-data))

This fix brought up the frame rate to **27-28 FPS**.

Second, it would be nice to have an asynchronous way to send the data. This could be done by replacing the existing submit method with following:

``` java
Transfer transfer = LibUsb.allocTransfer();
LibUsb.fillBulkTransfer(transfer, pushHandle, (byte) 0x01, headerBuffer, LibUsb::freeTransfer, null, 1000);

int result = LibUsb.submitTransfer(transfer);
if (result != LibUsb.SUCCESS) {
    throw new LibUsbException("Transfer of frame header to Push 2 display failed", result);
}
```

I'm currently working on this but it is not yet finished so I don't have an FPS speed up.

Third, is there a reason why to send the header and the data as two buffers? Wouldn't it be better to send them all together as one big chunk?  The documentation tells us to send two messages, but I'm not sure if it makes any difference for the device. Do you know something about this?

And thank you for developing this library, I was first doing it myself but then stumbled over your repo ;)

**Comment by cansik** [comment]
Ok I now just implemented the asynchronous sending of data and I am now on **50-60 FPS**. I think I have a big variation because of processing itself.

But there is still a problem when I shutdown the service that I get a java native error. I think it is because I am waiting not that long to before closing the connection to the device. But the speed is now perfect!

**Source Code**

``` java
/**
 * Send a frame of pixels async, corresponding to whatever has been drawn in the image returned by open(),
 * to the display.
 *
 * @throws LibUsbException       if there is a problem communicating.
 * @throws IllegalStateException if the Push 2 has not been opened.
 */
public static void sendFrameAsync() {
    if (transferBuffer == null) {
        throw new IllegalStateException("Push 2 device has not been opened");
    }
    IntBuffer transferred = IntBuffer.allocate(1);
    Transfer transfer = LibUsb.allocTransfer();
    LibUsb.fillBulkTransfer(transfer, pushHandle, (byte) 0x01, headerBuffer, LibUsb::freeTransfer, null, 1000);

    int result = LibUsb.submitTransfer(transfer);
    if (result != LibUsb.SUCCESS) {
        throw new LibUsbException("Transfer of frame header to Push 2 display failed", result);
    }

    // We send eight lines at a time to the display; allocate buffers big enough to receive them,
    // expand with the row stride padding, and mask with the signal shaping pattern.
    short[] pixels = new short[LINES_PER_TRANSFER * DISPLAY_WIDTH];
    byte[] maskedChunk = new byte[LINES_PER_TRANSFER * BYTES_PER_LINE];
    for (int i = 0; i < (DISPLAY_HEIGHT / LINES_PER_TRANSFER); i++) {
        displayImage.getRaster().getDataElements(
                0, i * LINES_PER_TRANSFER, DISPLAY_WIDTH, LINES_PER_TRANSFER, pixels);
        maskPixels(pixels, maskedChunk);
        transferBuffer.clear();
        transferBuffer.put(maskedChunk);
        transferred.clear();

        transfer = LibUsb.allocTransfer();
        LibUsb.fillBulkTransfer(transfer, pushHandle, (byte) 0x01, transferBuffer, LibUsb::freeTransfer, null, 1000);
        result = LibUsb.submitTransfer(transfer);

        if (result != LibUsb.SUCCESS) {
            throw new LibUsbException("Transfer of frame header to Push 2 display failed", result);
        }
    }
}
```

**Comment by cansik** [comment]
Ok the Fatal Error occurs when you close the device:

``` java
if (pushHandle != null) {
    displayImage = null;
    LibUsb.close(pushHandle);
    pushHandle = null;
}
```

I'm not sure if this really is necessary, I just commented it out. Because when you exit the context I think the device will also be closed.

**Comment by brunchboy** [comment]
Thanks so much for exploring these improvements, they sound very promising! I have just returned from nearly three weeks of vacationing in Europe and found that my computer had failed while I was away. It has been repaired and returned to me this morning, and I am in the process of restoring from my backups, so I cannot think in depth about Wayang right now, but should be able to this weekend. I hope that you will contribute a pull request with your speed improvements.

I would like to figure out why the close is crashing for you, though, and not for me. I don’t want to just comment it out because I need to be able to close and reopen the display without exiting the Java environment; I use Wayang in a lighting controller, and rely on being able to hot-plug and disconnect various hardware interfaces while the JVM environment and light and laser show stay up and running.

**Comment by cansik** [comment]
Yes it would be great if we could fix it. But I didn't found anything about the right way to close the usb4java device when you do asynchronous transfers. So I create a pull request when it's stable.

**Comment by cansik** [comment]
So just to document what I found out:

Earlier there was a problem that libusb didn't call the callback of the async method on OSX. That should be fixed now in the release `10.11.6`, but the callbacks are still not called.

But my tests showed that if I call `sendFrame()` (with `LibUsb.bulkTransfer(..)`), then libusb is calling my callbacks. Otherwise they won't be called and when you try to close the device, they are still waiting to be called.

So a simple workaround is, to sendFrame when you close the device. But maybe there is a simple call I can make to fix it.
